bus_procurement_cost/bus_cost_utils.py

```diff
@@ -0,0 +1,258 @@
+        return "other"
+
+    else:
+        return "not specified"
+    
+def project_type_finder(description: str) -> str:
+    """
+    function to match keywords to project description col to identify projects that only have bus procurement.
+    used to identify projects into diffferent categories: bus only, bus + others, no bus procurement.
+    use with .assign() to get a new col.
+    """
+    bus_list =[
+        "bus",
+        "transit vehicles",# for fta list
+        "cutaway vehicles",# for fta list
+        "zero-emission vehicles", # for tircp list
+        "zero emission vehicles",
+        "zero‐emissions vans",
+        "hybrid-electric vehicles",
+        "battery-electric vehicles",
+        "buy new replacement vehicles", # specific string for fta list
+    ]
+    
+    exclude_list =[
+        "facility",
+        #"station",
+        "stops",
+        "installation",
+        "depot",
+        "construct",
+        "infrastructure",
+        "signal priority",
+        "improvements",
+        "build",
+        "chargers",
+        "charging equipment",
+        "install",
+        "rail",
+        "garage",
+        "facilities",
+        "bus washing system",
+        "build a regional transit hub" # specific string needed for fta list
+        #"associated infrastructure" may need to look at what is associated infrastructure is for ZEB 
+        
+    ]
+    proj_description = description.lower().strip()
+
+    if any(word in proj_description for word in bus_list) and not any(
+        word in proj_description for word in exclude_list
+    ):
+        return "bus only"
+    
+    elif any(word in proj_description for word in exclude_list) and not any(
+        word in proj_description for word in bus_list
+    ):
+        return "non-bus components"
+    
+    elif any(word in proj_description for word in exclude_list) and any(
+        word in proj_description for word in bus_list
+    ):
+        return "includes bus and non-bus components"
+    
+    else:
+        return "needs review"
+    
+def col_row_updater(df: pd.DataFrame, col1: str, val1, col2: str, new_val):
+    """
+    function used to update values at specificed columns and row value.
+    """
+    df.loc[df[col1] == val1, col2] = new_val
+    
+    return
+
+#def bus_min_max_summary(data:pd.DataFrame, col1:str, col_list=["transit_agency",
+#                                                     "total_agg_cost",
+#                                                     "total_bus_count",
+#                                                     "new_cost_per_bus"]):
+#    """
+#    function to display min/max of specific column in aggregated bus df.
+    
+#    """
+    
+#    return display(Markdown(f"**Max {col1}**"),
+#        data[data[col1] == data[col1].max()][col_list],
+#        Markdown(f"**Min {col1}**"),
+#        data[data[col1] == data[col1].min()][col_list]
+#                  )
+
+def outlier_flag(col):
```

Comment:
```
def outlier_flag(df: pd.DataFrame, col: str) -> pd.DataFrame:
   
   # This applies the lambda function for you already, also worked in the absolute value
   # There are 2 ways to write this, to do the same thing
   
   return df.apply(lambda x: True if abs(x[col]) > 3 
   else False, axis=1)
   OR something like (double check)
   df[col].apply(lambda x: True if abs(x) > 3 else False) 
   
   # Also, if you don't like booleans, you can do `.astype(int)` and it'll change True/False or 1/0 (in that order)
```

In your notebook:

df_agg["new_is_cpb_outlier"] = outlier_flag(df_agg, "new_zscore_cost_per_bus")

Reply:
The `.apply` takes a lambda function, which operates on a row. There are 2 ways to write it, and this all depends on what you need to access in the row. If you have just 1 column (z-score), you can write it either way. If you need to access 2 column values, you will have to write it like `df.apply(lambda x: some condition, axis=1)`

Here, we want to access 2 columns in the lambda condition (`state` and `temperature`)
Ex: `df.apply(lambda x: 1 if (
(x.state == "CA" ) and (x.temperature  < 80) ) else 0, axis=1)`